### PR TITLE
Initial IPv6 Support

### PR DESCRIPTION
**Pull request by mnjdhl**
Initial code changes of IPv6 Support for CNDP

### CODE REVIEW

**Comment by KeithWiles** [comment]
I briefly looked at the diffs and notices a few formatting problem. Tabstop needs to be set to 4 and insert spaces for a tab.

Look at the CONTRIBUTING.md file and install the per-commit hook, which also requires clang-format to be installed. You will most likely have to edit the files and reformat replacing tabs with 4 spaces. A number of tools and editors will do this for you i.e,. vim, vscode, ....

I also make sure all of the new files have the correct SPDX license must be BSD-3-Clause. Also make sure the copyright dates reflect the correct year(s).

Plus so please add a 'Signed-off-by: Your Name <email-address>' to the commits.

The changes are looking very good!

**Comment by mnjdhl** [comment]
Thanks Keith for the feedback.
Will be following these guidelines.

**Comment by mnjdhl** [comment]
@KeithWiles 
The changes (formatting, copyright and some cleanup) are still underway.
Will be pushing them in a day or two.

**Comment by mnjdhl** [comment]
@KeithWiles 
Done the required changes. 
This commits contains:
- IPv6 protocol stack
- route for IPv6
- netif changes for IPv6

And still to come:
- NDP protocol and ICMP6 (minimum required)
- Integration of IPv6 with layer 2, layer 4 and FIB/RIB
- If anything more needed for IPv6 to work

Thanks for your support

**Comment by KeithWiles** [comment]
Please have a look at the CONTRIBUTING.md file it has some suggestion to add pre-commit hook and/or running clang-format on the code. If you have not already added the clang-format git hook.

**Comment by mnjdhl** [comment]
> Please have a look at the CONTRIBUTING.md file it has some suggestions like adding pre-commit hook and/or running clang-format on the code. If you have not already added the clang-format git hook.
Before pushing the commit 'git clang-format' was run. Apologize if you still see any formatting errors.

**Comment by KeithWiles** [comment]
Thank you. I was making sure it was run, but will look at the code when you are ready. If ready for a review please let me know.

**Comment by mnjdhl** [comment]
> Thank you. I was making sure it was run as the clang-format test was failing, but will look at the code when you are ready. If ready for a review please let me know.

Well, ipv6 functionality wise is not ready yet fully. This is work in progress. The integration with other layers is in progress and next will be  NDP/ICMP6. Likewise when all the dots are connected then only we can test the ipv6 functionality. 

So you can take a call whether to start review now or later when everything is ready.

**Comment by KeithWiles** [comment]
I can convert the PR to a draft (or you can) and then convert it back when ready. I would think we can wait until it is complete, but I will take a quick look at the code and provide some high level comments.

**Comment by mnjdhl** [comment]
> I can convert the PR to a draft (or you can) and then convert it back when ready. I would think we can wait until it is complete, but I will take a quick look at the code and provide some high level comments.

Have you looked at it?

**Comment by mnjdhl** [comment]
Thanks a lot Keith! 
Will start addressing the comments mostly after completing other parts.

File: lib/cnet/ipv6/ip6_flowlabel.h
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.h:4]: Done

File: lib/cnet/ipv6/ip6_flowlabel.h
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.h:42]: It is used in ip6_output.c as well

File: lib/cnet/ipv6/ip6_flowlabel.h
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.h:43]: Done

File: lib/cnet/ipv6/ip6_flowlabel.h
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.h:33]: Removed it

File: lib/cnet/ipv6/ip6_flowlabel.h
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.h:37]: Done it

File: lib/cnet/ipv6/ip6_flowlabel.h
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.h:43]: Done

And do you think we should move get_sysctl_value() to more generic source file?

**Comment by ekosi09** [comment]
Hello @KeithWiles 

To address your comments and queries about the code, @mnjdhl and I have made some changes and cleanups, so I would like to kindly ask you to review the PR again. 

It's rebased on top of the `main` CNDP project branch and has been evaluated by `clang-format` to follow the `CONTRIBUTING.md` guidelines.

Please let us know if there's any discrepancies or room for improvement.

Thank you in advance,
Emil

File: lib/cnet/route/cnet_route6.c
Comment (KeithWiles) [inline lib/cnet/route/cnet_route6.c:153]: Same as the above comment

File: usrtools/txgen/app/meson.build
Comment (KeithWiles) [inline usrtools/txgen/app/meson.build:23]: Can the file name be ipv6.c ?

File: usrtools/txgen/app/tg_ipv6.h
Comment (KeithWiles) [inline usrtools/txgen/app/tg_ipv6.h:5]: ipv6.h ??

**Review by KeithWiles — CHANGES_REQUESTED** [review]
Starting to look good. I only had time to do a quick review as too many files to review. Plus most of the comments are nit picks.

I was thinking we should add an option to meson_option.txt to enable/disable IPv6 support. This way we can commit this PR and not affect the current code base. By committing this PR we can have a smaller PR while continuing to improve or fix IPv6 support.
I was thinking disable IPv6 nodes, but allow the files to be compiled similar to the TCP code was done.

What do you think?

File: usrtools/txgen/app/tg_ipv6.h
Comment (mnjdhl) [reply usrtools/txgen/app/tg_ipv6.h:5]: Done

File: lib/cnet/chnl/cnet_chnl.c
Comment (mnjdhl) [reply lib/cnet/chnl/cnet_chnl.c:211]: IPPROTO_ICMP  is not implemented yet, that's why check is not there.

File: lib/cnet/chnl/cnet_chnl.c
Comment (mnjdhl) [reply lib/cnet/chnl/cnet_chnl.c:232]: IPPROTO_ICMP is not implemented yet, that's why check is not there.

File: lib/cnet/incs/cnet_fib_info.h
Comment (mnjdhl) [reply lib/cnet/incs/cnet_fib_info.h:380]: Done

File: lib/cnet/ipv6/ip6_flowlabel.h
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.h:18]: Done

File: lib/cnet/nd6/cnet_nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/cnet_nd6.c:69]: Done

File: lib/cnet/nd6/cnet_nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/cnet_nd6.c:89]: Done

File: lib/cnet/nd6/cnet_nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/cnet_nd6.c:129]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:25]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:17]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:48]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:118]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:185]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:194]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:204]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:223]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:341]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:363]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:477]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:529]: Done

File: usrtools/txgen/app/meson.build
Comment (mnjdhl) [reply usrtools/txgen/app/meson.build:23]: Done

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:313]: Done

File: lib/cnet/ipv6/ip6_flowlabel.c
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.c:17]: Done

File: lib/cnet/ipv6/ip6_flowlabel.c
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.c:27]: Done

File: lib/cnet/ipv6/ip6_flowlabel.c
Comment (mnjdhl) [reply lib/cnet/ipv6/ip6_flowlabel.c:91]: Done

File: lib/cnet/icmp6/icmp6_output.c
Comment (mnjdhl) [reply lib/cnet/icmp6/icmp6_output.c:49]: Removed it

File: lib/cnet/icmp6/icmp6_output.c
Comment (mnjdhl) [reply lib/cnet/icmp6/icmp6_output.c:51]: Removed it

File: lib/cnet/route/cnet_route6.c
Comment (mnjdhl) [reply lib/cnet/route/cnet_route6.c:147]: Removed the comment and added debug msg

File: lib/cnet/route/cnet_route6.c
Comment (mnjdhl) [reply lib/cnet/route/cnet_route6.c:153]: Removed the comment and added debug msg

File: lib/cnet/nd6/nd6.h
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.h:89]: Done

File: lib/cnet/ipv6/cnet_ipv6.h
Comment (mnjdhl) [reply lib/cnet/ipv6/cnet_ipv6.h:110]: Removed the comment now. If any issue comes we will add the extension later.

File: lib/cnet/nd6/nd6.c
Comment (mnjdhl) [reply lib/cnet/nd6/nd6.c:281]: I have left with a meaningful comment to fix this now.

**Comment by ekosi09** [comment]
Hello @KeithWiles 

We've cleaned up the code and @mnjdhl fixed the issues that you mentioned. Please look into these changes and give us feedback for further improvements.

Thank you in advance,
Emil

**Review by KeithWiles — COMMENTED** [review]
I did a quick review and it appears to be good. Thanks for making it optional to build.

Do you think we can merge this PR as is and then continue to debug IPv6 with new PRs?

**Comment by ekosi09** [comment]
Hello @KeithWiles 

Thank you for checking the changes. We would appreciate if you could point us in the right direction regarding the pre-commit hook issue since we do the `git clang-format --diff HEAD~1` and `git clang-format HEAD~1` manually to see and fix the formatting of the committed files and it's passing without errors. Only then we force-push the commit to the `feature-ipv6` branch.

In the `CONTRIBUTING.md` document, the hook and the manual `git clang-format` are described as the options for formatting along with the clang-format file `.clang-format` to use with ninja, however we don't do that option.

Also, I can see that there is already a `.git/hooks/pre-commit` hook located in the branch and it's executable:
```
$ ls -l .git/hooks/pre-commit*                                                                                                                  git:feature-ipv6
-rwxr-xr-x 1 ekosic ekosic  414 srp  19 13:14 .git/hooks/pre-commit
-rwxr-xr-x 1 ekosic ekosic 1643 srp  12 15:07 .git/hooks/pre-commit.sample
```

Thank you for your efforts,
Emil

**Comment by KeithWiles** [comment]
It appears you do have the pre-commit file setup correctly and here is the pre-commit file I have.

The pre-commit file should be run before you are able to commit the files. It maybe you need to modify the file in the feature-ipv6 branch by adding a space or space out a '=' to not be correct then commit the file. This way we can see if clang-format is working via the pre-commit hook.

foo = 1;
bar               = 2; // This line should be fixed by clang-format and align the '='

```
#! /bin/bash

for filename in $(git diff --cached --name-only | grep '.*\.[c|h]$'); do
  clang-format -style=file -i $filename; git add $filename;
done

for filename in $(git diff --cached --name-only | grep '.*\.go$'); do
  gofmt -s -w $filename; git add $filename;
done

for filename in $(git diff --cached --name-only | grep '.*\.jsonc$'); do
  ./tools/check_json.py --schema ./tools/cndp.schema $filename
done
```

**Comment by ekosi09** [comment]
Hello @KeithWiles 

Thank you for the explanation. I have tested the pre-commit hook just as you described and **I can confirm that it worked** (it fixed the clang-format issues right at committing the change):

1. I've made the following change (random variables):
```diff
diff --git a/lib/cnet/route/cnet_route6.c b/lib/cnet/route/cnet_route6.c
index 5359b86..9b3ec78 100644
--- a/lib/cnet/route/cnet_route6.c
+++ b/lib/cnet/route/cnet_route6.c
@@ -26,6 +26,9 @@
     ((1UL << RT6_NEXT_INDEX_SHIFT) - 1) /* MAX routes (16M) leaving bit 24-31 a next node index */
 #define RT6_DEFAULT_NUM_TBL8S (1 << 8)  /* Default number of tbl8 entries */
 
+uint16_t _test_a = 3;
+uint16_t _test_b            = 4;
+
```

2. I performed the addition and committing:
```
$ git commit -a -s -m 'testing pre-commit hook'
```

3. Then, I checked the file to see if clang-format was run:
```c
$ grep -C3 'uint16_t _test' lib/cnet/route/cnet_route6.c                                                                                        git:feature-ipv6
    ((1UL << RT6_NEXT_INDEX_SHIFT) - 1) /* MAX routes (16M) leaving bit 24-31 a next node index */
#define RT6_DEFAULT_NUM_TBL8S (1 << 8)  /* Default number of tbl8 entries */

uint16_t _test_a = 3;
uint16_t _test_b = 4;

int
cnet_route6_insert(int netdev_idx, struct in6_addr *dst, struct in6_addr *netmask,
```

This concludes that the pre-commit hook works as intended. Additionally to this, I've tried to run the hook script manually, and it does the same effect as just adding and committing the change, so I hope this concludes your concerns regarding the pre-commit hook.

As per your merging comment, I would like to kindly ask you if you could test the work since we didn't encounter any issues with it and we can then merge this commit and work on any additional fixes and issues later on.

Thank you in advance,
Emil

**Comment by KeithWiles** [comment]
I do not have time this week and I hope to have some time this weekend.

**Comment by KeithWiles** [comment]
I was working on reviewing and building the code this weekend. I ran into a number of warns and I tried to remove them, but I ran into more things I felt needed to be adjusted.

These changes caused a number of defines and structures to change, which affected a number of files. One example is I removed cne_fib6_conf structure and used cne_fib_conf everyplace. Then I moved CNE_FIB6_* to be CNE_FIB_* defines by moving them into the enum structures. Then I find that CNE_FIB6_IPV6_ADDR_SIZE could use the already defined IPV6_ADDR_LEN define. These changes and clang-format updating a number of files create a large number of file changes.

I wanted to give you a heads up on these changes and wanted to make sure these changes are OK. I would like to commit these changes, but wanted to make sure I was not stepping on anyone's toes and I would like to have you review them first of course.

I was not able to test the code other then running it, but I do not have the background to test IPv6 today.

Please let me know your thoughts and suggestions on how to move forward.

**Comment by KeithWiles** [comment]
I can try pushing my changes to the repo, if I have write permission.

**Comment by ekosi09** [comment]
Hello @KeithWiles 

You can try pushing your changes to the `feature-ipv6` branch on our side so we can review it.

Thank you,
Emil

**Comment by mnjdhl** [comment]
Hi Keith,

We have gone through your changes and  all of them look fine.

Please suggest the next step.

Thanks,

Manoj

**Comment by KeithWiles** [comment]
Hi Manoj,

I was thinking we could merge the code. This will allow you to start a new set of PR's to fix any bugs or add more features.
The current version of IPv6 is experimental and we need more testing time. Along with testing time it would be good to do some performance testing between IPv4 and IPv6.

Does that sound reasonable?

**Review by KeithWiles — APPROVED** [review]
Approving to allow more debugging and changes. The IPv6 is not enabled by default and is still in an experimental state.

**Comment by mnjdhl** [comment]
> Hi Manoj,
> 
> I was thinking we could merge the code. This will allow you to start a new set of PR's to fix any bugs or add more features. The current version of IPv6 is experimental and we need more testing time. Along with testing time it would be good to do some performance testing between IPv4 and IPv6.
> 
> Does that sound reasonable?

Sure Keith. Yeh it sounds good. Will merge it today.
Thanks a lot

**Comment by KeithWiles** [comment]
I am waiting on a response from another possible PR that will move a few directories around and break the repo up into more manageable repo's for distribution. If that is done before merging this PR, I will look into merging IPv6 into the new directory/repo structure.

**Comment by mnjdhl** [comment]
Okay Keith, we can wait for that PR.

**Comment by mnjdhl** [comment]
Thanks a ton Keith for merging the PR.
It's real pleasure to work with you.

Warm regards,

Manoj

**Pull request merged**
